examples/demo/src/orders/OrderList.tsx

```diff
@@ -0,0 +1,239 @@
+import {
+    AutocompleteInput,
+    BooleanField,
+    Datagrid,
+    DateField,
+    DateInput,
+    EditButton,
+    Filter,
+    List,
+    NullableBooleanInput,
+    NumberField,
+    ReferenceInput,
+    SearchInput,
+    TextField,
+    TextInput,
+} from 'react-admin';
+import {
+    makeStyles,
+    useMediaQuery,
+    Divider,
+    Tabs,
+    Tab,
+    Theme,
+} from '@material-ui/core';
+
+import NbItemsField from './NbItemsField';
+import CustomerReferenceField from '../visitors/CustomerReferenceField';
+import MobileGrid from './MobileGrid';
+import {
+    Customer,
+    FilterProps,
+    OrderStatus,
+    DatagridProps,
+    Order,
+    Props,
+    ListComponentProps,
+} from '../types';
+import { Identifier } from 'ra-core';
+
+interface FilterParams {
+    q?: string;
+    customer_id?: string;
+    date_gte?: string;
+    date_lte?: string;
+    total_gte?: string;
+    returned?: boolean;
+    status?: OrderStatus;
+}
+
+const OrderFilter: FC<FilterProps<FilterParams>> = props => (
+    <Filter {...props}>
+        <SearchInput source="q" alwaysOn />
+        <ReferenceInput source="customer_id" reference="customers">
+            <AutocompleteInput
+                optionText={(choice: Customer) =>
+                    choice.first_name && choice.last_name
+                        ? `${choice.first_name} ${choice.last_name}`
+                        : ''
+                }
+            />
+        </ReferenceInput>
+        <DateInput source="date_gte" />
+        <DateInput source="date_lte" />
+        <TextInput source="total_gte" />
+        <NullableBooleanInput source="returned" />
+    </Filter>
+);
+
+const useDatagridStyles = makeStyles({
+    total: { fontWeight: 'bold' },
+});
+
+const tabs = [
+    { id: 'ordered', name: 'ordered' },
+    { id: 'delivered', name: 'delivered' },
+    { id: 'cancelled', name: 'cancelled' },
+];
+
+interface TabbedDatagridProps extends DatagridProps<Order> {
+    classes: ReturnType<typeof useDatagridStyles>;
+    isXSmall: boolean;
+}
+
+const TabbedDatagrid: FC<TabbedDatagridProps> = ({
+    ids,
+    filterValues,
+    isXSmall,
+    classes,
+    setFilters,
+    displayedFilters,
+    ...rest
+}) => {
+    const [ordered, setOrdered] = useState<Identifier[]>([]);
+    const [delivered, setDelivered] = useState<Identifier[]>([]);
+    const [cancelled, setCancelled] = useState<Identifier[]>([]);
+
+    useEffect(() => {
+        if (ids && ids !== filterValues.status) {
+            switch (filterValues.status) {
+                case 'ordered':
+                    setOrdered(ids);
+                    break;
+                case 'delivered':
+                    setDelivered(ids);
+                    break;
+                case 'cancelled':
+                    setCancelled(ids);
+                    break;
+            }
+        }
+    }, [ids, filterValues.status]);
+
+    const handleChange = useCallback<
+        NonNullable<Props<typeof Tabs>['onChange']>
```

Comment:
cryptic type, please find a more readable way to do it

Reply:
Fixed